### PR TITLE
Minor performance improvements

### PR DESCRIPTION
**Pull request by amorde**

### CODE REVIEW

File: lib/cocoapods-core/dependency.rb [hunk 1]
```diff
@@ -109,11 +109,16 @@ def initialize(name = nil, *requirements)
     #         one or more version restrictions).
     #
     def requirement
-      if specific_version
-        Requirement.new(Version.new(specific_version.version))
-      else
-        @requirement
-      end
+      @specific_requirement || @requirement
```
Comment (amorde) [inline lib/cocoapods-core/dependency.rb:112]: `requirement` is accessed a large number of times during the dependency analysis phase. for cases when `specific_version` is non-nil, this becomes an expensive codepath

File: lib/cocoapods-core/dependency.rb [hunk 2]
```diff
@@ -196,9 +201,9 @@ def compatible?(other)
     def ==(other)
       self.class == other.class &&
         name == other.name &&
-        requirement == other.requirement &&
         external_source == other.external_source &&
-        podspec_repo == other.podspec_repo
+        podspec_repo == other.podspec_repo &&
+        requirement == other.requirement
```
Comment (amorde) [inline lib/cocoapods-core/dependency.rb:206]: `requirement` is a (relatively) heavier accessor than the others - short circuiting here improves runtime

File: lib/cocoapods-core/source.rb [hunk 3]
```diff
@@ -167,9 +167,10 @@ def versions(name)
       pod_dir = pod_path(name)
       return unless pod_dir.exist?
       @versions_by_name[name] ||= pod_dir.children.map do |v|
+        next nil unless v.directory?
```
Comment (amorde) [inline lib/cocoapods-core/source.rb:170]: avoid calling `v.basename.to_s` when `v.directory?` is `false`

File: lib/cocoapods-core/source.rb [hunk 4]
```diff
@@ -167,9 +167,10 @@ def versions(name)
...
         basename = v.basename.to_s
         begin
-          Version.new(basename) if v.directory? && basename[0, 1] != '.'
+          Version.new(basename) if basename[0, 1] != '.'
```
Comment (segiddins) [inline lib/cocoapods-core/source.rb:173]: nit: hoist into a `next nil unless`

File: lib/cocoapods-core/source.rb [hunk 5]
```diff
@@ -266,6 +267,11 @@ def search(query)
         query = query.root_name
       end
 
+      if (versions = @versions_by_name[query]) && !versions.empty?
```
Comment (amorde) [inline lib/cocoapods-core/source.rb:270]: if we already have a cached list of versions for this pod, skip globbing the pod directory to retrieve the list of versions

**Review by segiddins — APPROVED** [review]

**Review by dnkoutso — APPROVED** [review]

**Pull request merged**